### PR TITLE
Hdf5

### PR DESCRIPTION
**Pull request by scivision**
This uses pkg-config to find the HDF5 library directory, and patches the HDF5 standard filenames.

### CODE REVIEW

**Comment by jpakkane** [comment]
Master has been fixed. Please rebase so CI gets run. Thanks.

**Comment by scivision** [comment]
All the force pushes were to incorporate PRs that this PR relied on, and that were merged prior to this PR being merged. Now as you see it's just the original, single commit to add HDF5 to Meson.

**Comment by jpakkane** [comment]
All new features like this need test. I have updated the CI image to have `libhdf5-dev` installed. Please add a test and update `ciimage/Dockerfile` to add the package to the  install command (the CI image is updated manually due to reasons, but the change needs to go in the pull request).

File: mesonbuild/dependencies/misc.py [hunk 1]
```diff
@@ -33,6 +32,50 @@
...
+                    for larg in link_args:
+                        lpath = Path(larg)
+                        if lpath.is_file():                            
+                            if language == 'cpp':
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_hl_cpp' + lpath.suffix)))
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_cpp' + lpath.suffix)))
+                            elif language == 'fortran':
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + 'hl_fortran' + lpath.suffix)))
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_fortran' + lpath.suffix)))
+
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__ [W293] blank line contains whitespace

<sub>posted by Sider</sub>

File: mesonbuild/dependencies/misc.py [hunk 2]
```diff
@@ -33,6 +32,50 @@
...
+                        lpath = Path(larg)
+                        if lpath.is_file():                            
+                            if language == 'cpp':
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_hl_cpp' + lpath.suffix)))
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_cpp' + lpath.suffix)))
+                            elif language == 'fortran':
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + 'hl_fortran' + lpath.suffix)))
+                                lang_link_args.append(str(lpath.parent / (lpath.stem + '_fortran' + lpath.suffix)))
+                            
+                             # C is used by other languages
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__ [E114] indentation is not a multiple of four (comment)

<sub>posted by Sider</sub>

File: mesonbuild/dependencies/misc.py [hunk 3]
```diff
@@ -33,6 +32,50 @@
...
+            try:
+                pkgdep = PkgConfigDependency(pkg, environment, kwargs, language=self.language)
+                if pkgdep.found():
+                    self.compile_args = pkgdep.get_compile_args()
+                    # derive needed libraries by language
+                    link_args = pkgdep.get_link_args()
+                    lang_link_args = []
+                    for larg in link_args:
+                        lpath = Path(larg)
+                        if lpath.is_file():
```
Comment (ignatenkobrain) [inline mesonbuild/dependencies/misc.py]: __[Flake8]__ [W291] trailing whitespace

<sub>posted by Sider</sub>

**Comment by scivision** [comment]
@jpakkane tests are implemented and passing

**Comment by scivision** [comment]
I think it would be better to have the C++ and Fortran be optional also.
The HDF5 C library must be present for C++ and Fortran to work, but one might have HDF5 C without C++ or Fortran.

**Comment by scivision** [comment]
this breakage of #4860 is being handled in #4862

**Pull request merged**